File: packages/llm-test-harness-python/llm_test_harness/adapters/anthropic.py
from __future__ import annotations

import time
from datetime import datetime, timezone
from typing import Any

from ..cassette.match import find_interaction, require_interaction
from ..cassette.store import CassetteStore
from ..errors import CassetteOverwriteError
from ..types import (
    CassetteContentBlock,
    CassetteInteraction,
    CassetteMessage,
    CassetteMetadata,
    CassetteParams,
    CassetteRequest,
    CassetteResponse,
    CassetteUsage,
    ResolvedHarnessConfig,
)
def _build_request(params: dict[str, Any]) -> CassetteRequest:
    messages = [
        CassetteMessage(role=m["role"], content=m["content"])
        for m in params.get("messages", [])
    ]
    tools_raw = params.get("tools")
    tools = None
    if tools_raw:
        from ..types import CassetteTool
        tools = [
            CassetteTool(name=t.get("name", ""), description=t.get("description"))
            for t in tools_raw
        ]

    cp = CassetteParams(
        max_tokens=params.get("max_tokens"),
        temperature=params.get("temperature"),
        top_p=params.get("top_p"),
        stop=params.get("stop_sequences") or params.get("stop"),
    )
    return CassetteRequest(
        provider="anthropic",
        model=params.get("model", ""),
        system=params.get("system"),
        messages=messages,
        params=cp,
        tools=tools,
    )


def _response_to_cassette(resp: Any) -> CassetteResponse:
    content = None
    if hasattr(resp, "content") and resp.content:
        content = []
        for b in resp.content:
            block_type: str = getattr(b, "type", "text") or "text"
            text = getattr(b, "text", None)
            bid = getattr(b, "id", None)
            name = getattr(b, "name", None)
            inp = getattr(b, "input", None)
            content.append(
                CassetteContentBlock(type=block_type, text=text, id=bid, name=name, input=inp)
            )

    usage = None
    if hasattr(resp, "usage") and resp.usage:
        u = resp.usage
        usage = CassetteUsage(
            input_tokens=getattr(u, "input_tokens", None),
            output_tokens=getattr(u, "output_tokens", None),
        )

    return CassetteResponse(
        type=getattr(resp, "type", "message"),
        content=content,
        usage=usage,
        stop_reason=getattr(resp, "stop_reason", None),
    )


def _cassette_to_response(interaction: CassetteInteraction) -> Any:
    """Reconstruct a minimal Anthropic-like response object from cassette data."""

    class _Usage:
        def __init__(self, input_tokens: Any, output_tokens: Any) -> None:
            self.input_tokens = input_tokens
            self.output_tokens = output_tokens

    class _TextBlock:
        def __init__(self, text: str) -> None:
            self.type = "text"
            self.text = text

    class _ToolUseBlock:
        def __init__(self, bid: str, name: str, inp: Any) -> None:
            self.type = "tool_use"
            self.id = bid
            self.name = name
            self.input = inp

    class _Message:
        def __init__(
            self,
            content: list[Any],
            usage: Any,
            stop_reason: Any,
            model: str,
        ) -> None:
            self.type = "message"
            self.role = "assistant"
            self.content = content
            self.usage = usage
            self.stop_reason = stop_reason
            self.model = model
            self.id = f"replay_{interaction.id[:16]}"

    blocks: list[Any] = []
    if interaction.response.content:
        for b in interaction.response.content:
            if b.type == "tool_use":
                blocks.append(_ToolUseBlock(b.id or "", b.name or "", b.input))
            else:
                blocks.append(_TextBlock(b.text or ""))

    usage_obj = None
    if interaction.response.usage:
        usage_obj = _Usage(
            interaction.response.usage.input_tokens,
            interaction.response.usage.output_tokens,
        )

    return _Message(
        content=blocks,
        usage=usage_obj,
        stop_reason=interaction.response.stop_reason,
        model=interaction.request.model,
    )


def _intercept_create(
    original_create: Any,
    params: dict[str, Any],
    store: CassetteStore,
    config: ResolvedHarnessConfig,
) -> Any:
    # Streaming always passes through
    if params.get("stream"):
        return original_create(**params)

    req = _build_request(params)
    mode = config.mode

    if mode == "replay":
        interaction = require_interaction(store, req)
        return _cassette_to_response(interaction)

    if mode == "passthrough":
        return original_create(**params)

    if mode in ("record", "hybrid"):
        if mode == "hybrid":
            existing = find_interaction(store, req)
            if existing is not None:
                return _cassette_to_response(existing)

        start = time.monotonic()
        resp = original_create(**params)
        duration_ms = int((time.monotonic() - start) * 1000)

        from ..cassette.hash import hash_request
        hash_ = hash_request(req)

        if config.no_overwrite and store.find_by_id(hash_) is not None:
            raise CassetteOverwriteError(hash_)

        metadata = CassetteMetadata(
            recorded_at=datetime.now(timezone.utc).isoformat(),
            duration_ms=duration_ms,
            provider_request_id=getattr(resp, "id", None),
        )
        interaction = CassetteInteraction(
            id=hash_,
            request=req,
            response=_response_to_cassette(resp),
            metadata=metadata,
        )
        if config.on_before_record:
            interaction = config.on_before_record(interaction)
        store.append(interaction)
        return resp

    return original_create(**params)


class _MessagesProxy:
    def __init__(self, messages: Any, store: CassetteStore, config: ResolvedHarnessConfig) -> None:
        self._messages = messages
        self._store = store
        self._config = config

    def create(self, **params: Any) -> Any:
        return _intercept_create(self._messages.create, params, self._store, self._config)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._messages, name)


class AnthropicClientProxy:
    def __init__(self, client: Any, store: CassetteStore, config: ResolvedHarnessConfig) -> None:
        self._client = client
        self._store = store
        self._config = config

    @property
    def messages(self) -> _MessagesProxy:
        return _MessagesProxy(self._client.messages, self._store, self._config)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._client, name)

File: packages/llm-test-harness/src/adapters/anthropic.ts
import type { CassetteInteraction, CassetteRequest, ResolvedHarnessConfig } from '../types.js'
import type { CassetteStore } from '../cassette/cassette-store.js'
import { hashRequest } from '../cassette/cassette-hash.js'
import { findInteraction } from '../cassette/cassette-match.js'
import { CassetteMissError, CassetteOverwriteError } from '../errors.js'

// Minimal type definitions for Anthropic SDK shapes, kept local to avoid
// hard-coding the SDK version in the public API.

interface AnthropicMessageParam {
  role: 'user' | 'assistant'
  content: string | Array<{ type: string; text?: string; [key: string]: unknown }>
}

interface AnthropicCreateParams {
  model: string
  max_tokens: number
  messages: AnthropicMessageParam[]
  system?: string | Array<{ type: string; text: string }>
  temperature?: number
  top_p?: number
  stop_sequences?: string[]
  stream?: boolean
  tools?: Array<{ name: string; description?: string; input_schema: Record<string, unknown> }>
  [key: string]: unknown
}

interface AnthropicMessage {
  id: string
  type: 'message'
  role: 'assistant'
  content: Array<{ type: string; text?: string; [key: string]: unknown }>
  model: string
  stop_reason: string | null
  stop_sequence: string | null
  usage: { input_tokens: number; output_tokens: number }
}

function normalizeSystem(system: AnthropicCreateParams['system']): string | null {
  if (!system) return null
  if (typeof system === 'string') return system
  return system
    .filter((b) => b.type === 'text')
    .map((b) => b.text ?? '')
    .join('\n')
}

function normalizeContent(
  content: AnthropicMessageParam['content'],
): string {
  if (typeof content === 'string') return content
  return content
    .filter((b) => b.type === 'text')
    .map((b) => b.text ?? '')
    .join('\n')
}

function normalizeRequest(params: AnthropicCreateParams): CassetteRequest {
  return {
    provider: 'anthropic',
    model: params.model,
    system: normalizeSystem(params.system),
    messages: params.messages.map((m) => ({
      role: m.role,
      content: normalizeContent(m.content),
    })),
    params: {
      max_tokens: params.max_tokens ?? null,
      temperature: params.temperature ?? null,
      top_p: params.top_p ?? null,
      stop: params.stop_sequences ?? null,
    },
    tools: params.tools
      ? params.tools.map((t) => ({ name: t.name, description: t.description ?? null }))
      : null,
  }
}

function buildAnthropicResponse(
  interaction: CassetteInteraction,
  originalModel: string,
): AnthropicMessage {
  const { response, request, metadata } = interaction
  const content = (response.content ?? []).map((b) => {
    if (b.type === 'text') return { type: 'text' as const, text: b.text ?? '' }
    if (b.type === 'tool_use') {
      return { type: 'tool_use' as const, id: b.id ?? '', name: b.name ?? '', input: b.input ?? {} }
    }
    return { type: 'tool_result' as const }
  })

  return {
    id: metadata.provider_request_id ?? `replayed_${interaction.id.slice(0, 16)}`,
    type: 'message',
    role: 'assistant',
    content,
    model: request.model ?? originalModel,
    stop_reason: response.stop_reason ?? null,
    stop_sequence: null,
    usage: {
      input_tokens: response.usage?.input_tokens ?? 0,
      output_tokens: response.usage?.output_tokens ?? 0,
    },
  }
}

async function interceptCreate(
  params: AnthropicCreateParams,
  originalFn: (params: AnthropicCreateParams) => Promise<AnthropicMessage>,
  store: CassetteStore,
  config: ResolvedHarnessConfig,
): Promise<AnthropicMessage> {
  // Pass streaming through without cassette interception
  if (params.stream === true) {
    return originalFn(params)
  }

  const req = normalizeRequest(params)
  const hash = hashRequest(req)

  if (config.mode === 'replay' || config.mode === 'hybrid') {
    const interaction = findInteraction(store, req)
    if (interaction) {
      return buildAnthropicResponse(interaction, params.model)
    }
    if (config.mode === 'replay') {
      throw new CassetteMissError(hash)
    }
  }

  if (config.mode === 'passthrough') {
    return originalFn(params)
  }

  // record or hybrid-miss: call real API
  const start = Date.now()
  const response = await originalFn(params)
  const durationMs = Date.now() - start

  const interaction: CassetteInteraction = {
    id: hash,
    request: req,
    response: {
      type: 'message',
      content: response.content.map((b) => {
        if (b.type === 'text') return { type: 'text' as const, text: b.text ?? '' }
        if (b.type === 'tool_use') {
          return {
            type: 'tool_use' as const,
            id: (b as { id?: string }).id,
            name: (b as { name?: string }).name,
            input: (b as { input?: Record<string, unknown> }).input ?? {},
          }
        }
        return { type: 'tool_result' as const }
      }),
      usage: {
        input_tokens: response.usage.input_tokens,
        output_tokens: response.usage.output_tokens,
      },
      stop_reason: response.stop_reason ?? undefined,
    },
    metadata: {
      recorded_at: new Date().toISOString(),
      duration_ms: durationMs,
      provider_request_id: response.id,
    },
  }

  const final = config.onBeforeRecord(interaction)

  if (config.noOverwrite && store.findById(hash) !== undefined) {
    throw new CassetteOverwriteError(hash)
  }

  store.append(final)
  return response
}

export function wrapAnthropicClient<T extends object>(
  client: T,
  store: CassetteStore,
  config: ResolvedHarnessConfig,
): T {
  return new Proxy(client, {
    get(target, prop, receiver) {
      if (prop === 'messages') {
        const originalMessages = Reflect.get(target, prop, target) as Record<string, unknown>
        return new Proxy(originalMessages, {
          get(msgTarget, msgProp) {
            if (msgProp === 'create') {
              return (params: AnthropicCreateParams) =>
                interceptCreate(
                  params,
                  (p) =>
                    (msgTarget['create'] as (p: AnthropicCreateParams) => Promise<AnthropicMessage>).call(
                      msgTarget,
                      p,
                    ),
                  store,
                  config,
                )
            }
            return Reflect.get(msgTarget, msgProp, msgTarget)
          },
        })
      }
      return Reflect.get(target, prop, receiver)
    },
  })
}
